Source/WebKit/WPEPlatform/wpe/headless/WPEDisplayHeadless.cpp

```
@@ -68,16 +91,63 @@ static WPEView* wpeDisplayHeadlessCreateView(WPEDisplay* display)
-    wpe_buffer_dma_buf_formats_builder_append_format(builder, DRM_FORMAT_ARGB8888, DRM_FORMAT_MOD_LINEAR);
-    return wpe_buffer_dma_buf_formats_builder_end(builder);
+#if USE(GBM)
+    if (const char* filename = wpe_display_get_drm_render_node(display)) {
+        if (!epoxy_has_egl_extension(nullptr, "EGL_KHR_platform_gbm")) {
+            g_set_error_literal(error, WPE_EGL_ERROR, WPE_EGL_ERROR_NOT_AVAILABLE, "Can't get EGL display: GBM platform not supported");
+            return nullptr;
+        }
+
+        auto fd = UnixFileDescriptor { open(filename, O_RDWR | O_CLOEXEC), UnixFileDescriptor::Adopt };
+        if (!fd) {
+            g_set_error(error, WPE_EGL_ERROR, WPE_EGL_ERROR_NOT_AVAILABLE, "Can't get EGL display: failed to open device %s", filename);
+            return nullptr;
+        }
+        auto* device = gbm_create_device(fd.value());
+        if (!device) {
+            g_set_error(error, WPE_EGL_ERROR, WPE_EGL_ERROR_NOT_AVAILABLE, "Can't get EGL display: failed to create GBM device for %s", filename);
+            return nullptr;
+        }
+
+        EGLDisplay eglDisplay = EGL_NO_DISPLAY;
+        if (epoxy_has_egl_extension(nullptr, "EGL_EXT_platform_base"))
+            eglDisplay = eglGetPlatformDisplayEXT(EGL_PLATFORM_GBM_KHR, device, nullptr);
+        else if (epoxy_has_egl_extension(nullptr, "EGL_KHR_platform_base"))
+            eglDisplay = eglGetPlatformDisplay(EGL_PLATFORM_GBM_KHR, device, nullptr);
+
+        if (eglDisplay != EGL_NO_DISPLAY) {
+            auto* priv = WPE_DISPLAY_HEADLESS(display)->priv;
+            priv->gbmDeviceFD = WTFMove(fd);
+            priv->gbmDevice = device;
+            return eglDisplay;
+        }
+
+        gbm_device_destroy(device);
+        g_set_error(error, WPE_EGL_ERROR, WPE_EGL_ERROR_NOT_AVAILABLE, "Can't get EGL display: failed to create GBM EGL display for %s", filename);
+        return nullptr;
```

Shouldn't this fall-through instead of returning, and try to use the Mesa surfaceless platform as fallback?

---

No, if we have a device we should use GBM, because we can't ensure the device used by the surfaceless platform will be the one we have

---

Good point!